docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,270 @@
+* In the `awaiting_sub_operation` state, workflow just waits monitoring the state of the sub-operation completion.
+  Once the sub-operation is successful, the workflow moves back to the `executing` state with the index value incremented,
+  so that the next item can be applied.
+  * In case of a failure, the workflow moves to the `on_error` target state, keeping the `current_index` value intact,
+    so that the item that caused the failure can be easily identified.
+* Once the `updated_profile` list is exhausted in the `executing` state, the workflow moves to its `on_success` target state.
+
+
+### On success
+
+Once all the operations in the profile are successfully completed, the successful status is published
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "successful",
+  ... // Other input fields
+}'
+```
+
+...and the current applied device profile information is updated by publishing the same to the capability topic as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile' '{
+  "name": "prod-profile",
+  "version: "v2"
+}'
+```
+
+### On failure
+
+On failure, the `device_profile` operation is aborted at the operation that caused the failure.
+The remaining operations in the profile are executed and no attempt is made to rollback the already completed operations either.
+If the sub-operations support rollbacks at the sub-operation level, it is performed for the failed operation.
+
+For example, if a profile includes a firmware, 2 software and 1 configuration updates in that sequence,
+if the failure happens during the second software update, no rollback is performed at the overall `device_profile` operation level,
+or even for that failed software update, unless the `software_update` workflow for that software type supports rollbacks.
+In that case the firmware and 1st software would remain installed, with the failed software update and last config update skipped.
+But, if the failure happens during the `firmware_update` itself, a rollback is most likely performed by that workflow,
+as most `firmware_update` workflows support a robust rollback mechanism.
+
+The `device_profile` operation itself is marked failed as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "failed",
+  "reason": "Installation of software module: `jq` failed. Refer to operation logs for further details."
+  ... // Other input fields
+}'
+```
+
+The same profile can be applied again after fixing the issues that caused the failure,
+and it is left to the individual sub-operations to determine whether
+the same operation that was successfully applied in the last attempt must be reapplied or not.
+The user can also manually skip any operation using the `skip` field.
+
+# Cumulocity operation mapping
+
+TBD
```

Comment:
The main question to address is "how to define the order in which the operations have to be executed, while c8y just provides an unordered set of actions".

An option is to let the cloud user include in the device profile set a script with a well-know name that compute the desire order. The device profile workflow would look for that script in the device profile and would update its own config first thing if found.

Reply:
The Cumulocity IoT mapping shouldn't matter so much as the sorting belongs in the workflow scheduled step. So we can just take the list as it stands and initially sort it firmware, software, configuration (and keeping any sorting of individual sections)